File: prolog/src/movement.pl
/* movement.pl */
:- multifile i_am_at/1, at/2, holding/1, path/3, door_unlocked/1, interact/2, describe/1, escape_city_ending/0, hill_church_ending_escape/0.

/* State of rooms */
door_unlocked(hotel_room) :- fail.
door_unlocked(hotel_basement) :- fail.
door_unlocked(archive) :- fail.

/* This rule tells how to look around you. */
look :-
    i_am_at(Place),
    describe(Place),
    notice_objects_at(Place),
    nl.


/* These rules set up a loop to mention all the objects
in your vicinity. */

notice_objects_at(Place) :-
    at(X, Place),
    write('There is '), write(X), write(' here.'), nl,
    fail.

notice_objects_at(_).

/* These rules define the direction letters as calls to go/1. */

n :- go(n).

s :- go(s).

e :- go(e).

w :- go(w).


/* These rules tells how to move in a given direction. */

/* Conditional rules for moving */
go(e) :-
        i_am_at(train_station),
        escape_city_ending.
go(w) :-
    i_am_at(hotel_corridor),
    \+ door_unlocked(hotel_room),
        write('Door is locked.'), nl,
        write('Next to the door, there is a sign that reads: "Michael Turner".'), nl,
        write('You have to enter 4 digit code to unlock the door.'), nl,
        read(Code),
        try_unlock_hotel_room(Code),
        !.

go(w) :-
    i_am_at(hotel_lobby),
    \+ door_unlocked(hotel_basement),
    try_unlock_hotel_basement,
    !.

go(n) :-
    i_am_at(hill_church_second_floor),
    holding(weird_box),
    hill_church_ending_escape.

go(n) :-
    i_am_at(forest_cave),
    \+ holding(amulet),
    write('A strange force seems to block your path. The entrance won’t budge.'), nl,
    write('It feels like something is missing, something that could unlock the cave\'s secrets.'), nl,
    !.

go(n) :-
    i_am_at(forest_cave),
    holding(amulet),
    write('The amulet in your possession glows faintly, and you feel an ancient force giving way.'), nl,
    write('The entrance opens, allowing you to pass into the unknown.'), nl,
    retract(i_am_at(forest_cave)),
    assert(i_am_at(ending_cave)),
    !, look.

go(s) :-
    i_am_at(library),
    \+ door_unlocked(archive),
        ( holding(key) ->
            write('You open the door with the key.'), nl,
            retractall(holding(key)),
            assert(door_unlocked(archive)),
            go(s)
        ;
            write('The door is locked.'), nl,
            write('You need a key to unlock it.'), nl
        ).

/* General rules for moving */
go(Direction) :-
    i_am_at(Here),
    path(Here, Direction, There),
    retract(i_am_at(Here)),
    assert(i_am_at(There)),
    !, look.

go(_) :-
    write('You can''t go that way.').

/* Rules for entering and exiting the car */

enter(car) :-
    i_am_at(parking),
    holding(car_keys),
    write('You unlock the car with the keys.'), nl,
    retract(i_am_at(parking)),
    assert(i_am_at(car)),
    look.

enter(car) :-
    i_am_at(parking),
    \+ holding(car_keys),
    write('You need the car keys to enter the car.'), nl.

enter(car) :-
    i_am_at(car),
    write('You are already inside the car.'), nl.

enter(car) :-
    write('There is no car here.'), nl.

exit(car) :-
    i_am_at(car),
    write('You get out of the car and return to the parking lot.'), nl,
    retract(i_am_at(car)),
    assert(i_am_at(parking)).

/* This rule tells how to unlock the hotel room door. */

try_unlock_hotel_room(Code) :-
    /* The code can be found in a different location as year of birth of Michael Turner */
    number(Code),
    Code =:= 1974,
    retractall(door_unlocked(hotel_room)),
    assert(door_unlocked(hotel_room)),
    nl,
    write('You have unlocked the door.'), nl,
    go(w),
    !.

try_unlock_hotel_room(_) :-
    write('Wrong code.'), nl,
    !, look.

/* This rule tells how to unlock the hotel basement door. */

try_unlock_hotel_basement :-
    (holding(red_fuse), holding(blue_fuse), holding(green_fuse) ->
        nl,
        write('You put all 3 fuses in the fuse box.'), nl,
        write('The elevator starts working.'), nl,
        retractall(holding(red_fuse)),
        retractall(holding(blue_fuse)),
        retractall(holding(green_fuse)),
        retractall(door_unlocked(hotel_basement)),
        assert(door_unlocked(hotel_basement)),
        go(w)
    ;
        nl,
        write('The elevator is not working.'), nl,
        write('However, you notice a fuse box next to the elevator with 3 fuses missing.'), nl,
        nl,
        write('"Maybe if I find the fuses, I can get the elevator working."'), nl,
        look
    ).

/* This rule tells how to unlock the safe. */

open_safe(Code) :-
    /* The code can be found as an easter egg in the leaf code drunkard gave you */
    i_am_at(police_station),
    number(Code),
    Code =:= 2137,
    assert(holding(engraved_ring)),
    nl,
    write('The safe opens, revealing an engraved ring inside.'), nl,
    write('You take the engraved ring.'), nl,
    !.

open_safe(_) :-
    write('Wrong code.'), nl,
    !, look.

File: haskell/src/Movement.hs
module Movement where

import Control.Monad.State (MonadState (put))
import Data.Map qualified as Map
import Endings
import GameState
import LocationUtils
import Locations
import Look
import Objects

move :: String -> String -> GameState -> IO GameState
move "east" "train_station" state = do
  putStrLn "Are you sure you want to escape the city? (yes/no)"
  confirmation <- getLine
  if confirmation == "yes"
    then do
      let message = escapeCityEndingText (hasItem "diary" state)
      putStrLn message
      let updatedState = state {gameOver = True}
      return updatedState
    else do
      lookAround state
move "north" "second_floor_of_hill_church" state =
  if "weird_box_opened" `elem` flags state
    then do
      let message = hillChurchEndingEscapeText
      putStrLn message
      let updatedState =
            state
              { gameOver = True
              }
      return updatedState
    else do
      putStrLn "You can't go that way"
      return state
move "north" "forest_cave" state = do
  if hasItem "amulet" state
    then do
      putStrLn "The amulet in your possession glows faintly, and you feel an ancient force giving way."
      putStrLn "The entrance opens, allowing you to pass into the unknown."
      if hasItem "engraved_ring" state
        then do
          let message = forestCaveEndingWeakenedText
          putStrLn message
        else do
          let message = forestCaveEndingKilledText
          putStrLn message
      let updatedState =
            state
              { gameOver = True
              }
      return updatedState
    else do
      putStrLn "A strange force seems to block your path. The entrance won’t budge."
      putStrLn "It feels like something is missing, something that could unlock the cave\'s secrets."
      return state
move strDirection currentLocationStr state =
  case parseDirection strDirection of
    Nothing ->
      putStrLn "Invalid direction" >> return state
    Just direction ->
      case Map.lookup direction (directions (currentLocation state)) of
        Just (nextLocationName, True) -> goNextLocation nextLocationName state
        Just (nextLocationName, False) -> locationBlocked nextLocationName state
        Nothing ->
          putStrLn "You can't go that way" >> return state

goNextLocation nextLocationName state =
  case findLocation nextLocationName of
    Just nextLocation ->
      do
        let updatedState = state {currentLocation = nextLocation}
        lookAround updatedState
    Nothing ->
      do
        putStrLn "Error: Location not found!"
        return state

locationBlocked :: String -> GameState -> IO GameState
locationBlocked nextLocationName state =
  if nextLocationName `elem` closedLocations state
    then case nextLocationName of
      "hotel_room" ->
        openHotelRoom state
      "hotel_basement" ->
        openHotelBasement state
      "archive" ->
        openArchive state
      _ ->
        putStrLn "Error: Don't know how to open it" >> return state
    else
      goNextLocation nextLocationName state

openHotelRoom :: GameState -> IO GameState
openHotelRoom state = do
  putStrLn "The door is locked."
  putStrLn "However, you noticed a sign that says 'Michael Turner'."
  putStrLn "There is also a keypad next to the door."
  putStrLn "Please enter 4-digit code:"
  code <- getLine
  if code == "1974"
    then do
      let updatedState =
            state
              { closedLocations = filter (/= "hotel_room") (closedLocations state)
              }
      putStrLn "Correct code!"
      goNextLocation "hotel_room" updatedState
    else do
      putStrLn "Wrong code!"
      return state

openHotelBasement :: GameState -> IO GameState
openHotelBasement state =
  if redFuse `elem` inventory state
    && blueFuse `elem` inventory state
    && greenFuse `elem` inventory state
    then do
      let updatedState =
            state
              { inventory = filter (`notElem` [redFuse, greenFuse, blueFuse]) (inventory state),
                closedLocations = filter (/= "hotel_basement") (closedLocations state)
              }
      putStrLn "You placed the fuses in the fuse box and the elevator started working."
      goNextLocation "hotel_basement" updatedState
    else do
      putStrLn "The elevator is not working."
      putStrLn "However, you noticed a fuse box next to the elevator with 3 fuses missing."
      putStrLn "Maybe you can find them somewhere."
      return state

openArchive :: GameState -> IO GameState
openArchive state =
  if key `elem` inventory state
    then do
      let updatedState =
            state
              { inventory = filter (/= key) (inventory state),
                closedLocations = filter (/= "archive") (closedLocations state)
              }
      putStrLn "You have unlocked the archive."
      goNextLocation "archive" updatedState
    else do
      putStrLn "The archive is locked. You need a key to open it."
      return state
